File: iosApp/deadly/Core/Database/DAOs/PlayQueueDAO.swift
import Foundation
import GRDB

/// Data access for the persistent show queue (ADR-0010).
///
/// Ordering is by `position` ascending; the head (next show) is the lowest
/// position. Append goes to MAX(position)+1; head-insert shifts all rows down
/// and inserts at 0.
struct PlayQueueDAO: Sendable {
    let database: AppDatabase

    // MARK: - Fetch

    func fetchAll() throws -> [PlayQueueRecord] {
        try database.read { db in
            try PlayQueueRecord.order(Column("position").asc).fetchAll(db)
        }
    }

    func peekHead() throws -> PlayQueueRecord? {
        try database.read { db in
            try PlayQueueRecord.order(Column("position").asc).fetchOne(db)
        }
    }

    func contains(showId: String) throws -> Bool {
        try database.read { db in
            try PlayQueueRecord.filter(Column("showId") == showId).fetchCount(db) > 0
        }
    }

    // MARK: - Mutate

    /// Append to the tail.
    func append(showId: String, recordingId: String?) throws {
        try database.write { db in
            let maxPos = try Int.fetchOne(db, sql: "SELECT MAX(position) FROM play_queue") ?? -1
            var rec = PlayQueueRecord(
                id: nil, showId: showId, recordingId: recordingId,
                position: maxPos + 1, resumeTrackIndex: nil, resumePositionMs: nil,
                addedAt: Int64(Date().timeIntervalSince1970 * 1000)
            )
            try rec.insert(db)
        }
    }

    /// Insert at the head (position 0), shifting existing rows down. Carries an
    /// optional resume snapshot (interrupt re-queue).
    func insertHead(showId: String, recordingId: String?, resumeTrackIndex: Int?, resumePositionMs: Int64?) throws {
        try database.write { db in
            try db.execute(sql: "UPDATE play_queue SET position = position + 1")
            var rec = PlayQueueRecord(
                id: nil, showId: showId, recordingId: recordingId,
                position: 0, resumeTrackIndex: resumeTrackIndex, resumePositionMs: resumePositionMs,
                addedAt: Int64(Date().timeIntervalSince1970 * 1000)
            )
            try rec.insert(db)
        }
    }

    func delete(id: Int64) throws {
        _ = try database.write { db in
            try PlayQueueRecord.deleteOne(db, key: id)
        }
    }

    /// Remove and return the head (end-of-show auto-advance).
    func popHead() throws -> PlayQueueRecord? {
        try database.write { db in
            guard let head = try PlayQueueRecord.order(Column("position").asc).fetchOne(db) else { return nil }
            if let id = head.id { _ = try PlayQueueRecord.deleteOne(db, key: id) }
            return head
        }
    }

    func clear() throws {
        _ = try database.write { db in
            try PlayQueueRecord.deleteAll(db)
        }
    }

    /// Rewrite positions to match the given ordered ids (index becomes position).
    func reorder(orderedIds: [Int64]) throws {
        try database.write { db in
            for (index, id) in orderedIds.enumerated() {
                try db.execute(sql: "UPDATE play_queue SET position = ? WHERE id = ?", arguments: [index, id])
            }
        }
    }
}

File: androidApp/core/database/src/main/java/com/grateful/deadly/core/database/dao/PlayQueueDao.kt
package com.grateful.deadly.core.database.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.Query
import androidx.room.Transaction
import com.grateful.deadly.core.database.entities.PlayQueueEntity
import kotlinx.coroutines.flow.Flow

/**
 * Data access for the persistent show queue (ADR-0010).
 *
 * Ordering is by [PlayQueueEntity.position] ascending; the head (next show) is
 * the lowest position. Append goes to [maxPosition] + 1; head-insert shifts all
 * rows down and inserts at 0.
 */
@Dao
interface PlayQueueDao {

    @Query("SELECT * FROM play_queue ORDER BY position ASC")
    fun observeQueue(): Flow<List<PlayQueueEntity>>

    @Query("SELECT * FROM play_queue ORDER BY position ASC")
    suspend fun getQueue(): List<PlayQueueEntity>

    @Query("SELECT * FROM play_queue ORDER BY position ASC LIMIT 1")
    suspend fun peekHead(): PlayQueueEntity?

    @Query("SELECT COALESCE(MAX(position), -1) FROM play_queue")
    suspend fun maxPosition(): Int

    @Query("SELECT EXISTS(SELECT 1 FROM play_queue WHERE showId = :showId)")
    suspend fun containsShow(showId: String): Boolean

    @Insert
    suspend fun insert(entity: PlayQueueEntity): Long

    @Query("UPDATE play_queue SET position = position + 1")
    suspend fun shiftAllDown()

    @Query("DELETE FROM play_queue WHERE id = :id")
    suspend fun deleteById(id: Long)

    /** Remove all entries for a show. Used when a show becomes current — playing
     *  any show pops it from the upcoming queue (ADR-0010 §1). */
    @Query("DELETE FROM play_queue WHERE showId = :showId")
    suspend fun deleteByShowId(showId: String)

    @Query("DELETE FROM play_queue")
    suspend fun clear()

    @Query("UPDATE play_queue SET position = :position WHERE id = :id")
    suspend fun updatePosition(id: Long, position: Int)

    /** Rewrite positions to match [orderedIds]; index in the list becomes position. */
    @Transaction
    suspend fun reorder(orderedIds: List<Long>) {
        orderedIds.forEachIndexed { index, id -> updatePosition(id, index) }
    }

    /** Append [entity] at the tail in one round-trip. */
    @Transaction
    suspend fun append(entity: PlayQueueEntity): Long {
        val next = maxPosition() + 1
        return insert(entity.copy(position = next))
    }

    /** Insert [entity] at the head (position 0), shifting existing rows down. */
    @Transaction
    suspend fun insertHead(entity: PlayQueueEntity): Long {
        shiftAllDown()
        return insert(entity.copy(position = 0))
    }
}
